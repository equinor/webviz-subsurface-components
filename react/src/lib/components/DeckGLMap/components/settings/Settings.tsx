import { Icon } from "@equinor/eds-core-react";
import { layers } from "@equinor/eds-icons";
import { createStyles, makeStyles, Theme } from "@material-ui/core";
import { isEmpty } from "lodash";
import React from "react";
import { useSelector } from "react-redux";
import { MapState } from "../../redux/store";
import LayersButton from "./LayersButton";
import LayerSettingsButton from "./LayerSettingsButton";

Icon.add({ layers }); // (this needs only be done once)

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: "absolute",
            bottom: theme.spacing(2),
            right: theme.spacing(2),
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
        },
    })
);
const Settings: React.FC = React.memo(() => {
    const classes = useStyles();

    const spec = useSelector((st: MapState) => st.spec);

    if (isEmpty(spec) || !("layers" in spec)) return null;
    return (
        <div className={classes.root}>
            {/* eslint-disable-next-line @typescript-eslint/no-explicit-any */}
            {(spec["layers"] as any[]).map((layer) => (
                <LayerSettingsButton
                    layerId={layer.id}
                    layerType={layer["@@type"]}
                    key={`settings-button-${layer.id}`}
                />
            ))}
            <LayersButton />
        </div>
    );
});

Settings.displayName = "Settings";
export default Settings;
