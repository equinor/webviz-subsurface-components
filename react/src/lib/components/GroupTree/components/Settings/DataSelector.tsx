import { NativeSelect } from "@equinor/eds-core-react";
import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React, { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { updateCurrentDataType } from "../../redux/actions";
import { GroupTreeState } from "../../redux/store";
import { DataType, DataTypes } from "../../redux/types";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            minWidth: "170px",
            maxWidth: "170px",
            padding: theme.spacing(1),
        },
    })
);

const DataSelector: React.FC = React.memo(() => {
    const classes = useStyles();
    // Redux
    const dispatch = useDispatch();
    const currentDataType = useSelector(
        (st: GroupTreeState) => st.ui.currentDataType
    );
    // handlers
    const handleSelectedItemChange = useCallback(
        (event) => dispatch(updateCurrentDataType(event.target.value)),
        [dispatch]
    );

    return (
        <NativeSelect
            className={classes.root}
            id="data-type-selector"
            label="Data Type"
            value={currentDataType}
            onChange={handleSelectedItemChange}
        >
            {Object.keys(DataTypes).map((datatype) => (
                <option key={`option-${datatype}`} value={datatype}>
                    {DataTypes[datatype as DataType]}
                </option>
            ))}
        </NativeSelect>
    );
});

DataSelector.displayName = "DataSelector";
export default DataSelector;
