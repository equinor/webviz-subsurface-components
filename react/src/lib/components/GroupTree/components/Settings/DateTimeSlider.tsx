import {
    createStyles,
    makeStyles,
    Slider,
    Theme,
    // eslint-disable-next-line prettier/prettier
    withStyles
} from "@material-ui/core";
import React, { useCallback, useContext, useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { updateCurrentDateTime } from "../../redux/actions";
import { GroupTreeState } from "../../redux/store";
import { Data, DatedTree } from "../../redux/types";
import { DataContext } from "../DataLoader";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            width: "200px",
            marginRight: theme.spacing(4),
        },
    })
);
const EdsSlider = withStyles({
    root: {
        color: "#007079",
    },
    valueLabel: {
        top: 22,
        "& *": {
            background: "transparent",
            color: "#000",
        },
    },
})(Slider);
const DateTimeSlider: React.FC = React.memo(() => {
    const classes = useStyles();
    const data: Data = useContext(DataContext);
    // Redux
    const dispatch = useDispatch();

    const currentDateTime = useSelector(
        (state: GroupTreeState) => state.ui.currentDateTime
    );

    const times = useMemo(
        // list of all dates
        () => {
            const times: string[] = data.reduce(
                (total: string[], currentValue: DatedTree) => {
                    return total.concat(currentValue.dates);
                },
                []
            );
            return times;
        },
        [data]
    );

    const currentDateTimeIndex = useMemo(
        () => times.indexOf(currentDateTime),
        [times, currentDateTime]
    );

    // handlers
    const outputFunction = useCallback((step: number) => times[step], [times]);
    const onChange = useCallback(
        (_, step) => dispatch(updateCurrentDateTime(times[step])),
        [dispatch, times]
    );
    return (
        <div className={classes.root}>
            <span>Time Steps</span>
            <EdsSlider
                track={false}
                aria-labelledby="date-time-slider-label"
                value={currentDateTimeIndex}
                valueLabelDisplay="on"
                onChange={onChange}
                min={0}
                max={times.length - 1}
                step={1}
                marks={true}
                valueLabelFormat={outputFunction}
            />
        </div>
    );
});

DateTimeSlider.displayName = "DateTimeSlider";
export default DateTimeSlider;
