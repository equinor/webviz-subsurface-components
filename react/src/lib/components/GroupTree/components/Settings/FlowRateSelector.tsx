import { NativeSelect } from "@equinor/eds-core-react";
import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React, { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { updateCurrentFlowRate } from "../../redux/actions";
import { GroupTreeState } from "../../redux/store";
import { DataInfos, DataInfo } from "../../redux/types";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            maxWidth: "250px",
            padding: theme.spacing(1),
        },
    })
);

interface Props {
    edge_options: DataInfos;
}

const FlowRateSelector: React.FC<Props> = React.memo(
    ({ edge_options }: Props) => {
        const classes = useStyles();
        // Redux
        const dispatch = useDispatch();
        const currentFlowRate = useSelector(
            (st: GroupTreeState) => st.ui.currentFlowRate
        );
        // handlers
        const handleSelectedItemChange = useCallback(
            (event) => {
                dispatch(updateCurrentFlowRate(event.target.value));
            },
            [dispatch]
        );

        return (
            <NativeSelect
                className={classes.root}
                id="flow-rate-selector"
                label="Flow Rate"
                value={currentFlowRate}
                onChange={handleSelectedItemChange}
            >
                {edge_options.map((key: DataInfo) => (
                    <option key={`option-${key.name}`} value={key.name}>
                        {key.label}
                    </option>
                ))}
            </NativeSelect>
        );
    }
);

FlowRateSelector.displayName = "FlowRateSelector";
export default FlowRateSelector;
