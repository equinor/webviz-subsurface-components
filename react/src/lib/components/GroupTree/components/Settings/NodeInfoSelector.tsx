import { NativeSelect } from "@equinor/eds-core-react";
import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React, { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { updateCurrentNodeInfo } from "../../redux/actions";
import { GroupTreeState } from "../../redux/store";
import { DataInfos, DataInfo } from "../../redux/types";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            maxWidth: "250px",
            padding: theme.spacing(1),
        },
    })
);

interface Props {
    node_options: DataInfos;
}

const NodeInfoSelector: React.FC<Props> = React.memo(
    ({ node_options }: Props) => {
        const classes = useStyles();
        // Redux
        const dispatch = useDispatch();
        const currentNodeInfo = useSelector(
            (st: GroupTreeState) => st.ui.currentNodeInfo
        );
        // handlers
        const handleSelectedItemChange = useCallback(
            (event) => {
                dispatch(updateCurrentNodeInfo(event.target.value));
            },
            [dispatch]
        );

        return (
            <NativeSelect
                className={classes.root}
                id="node-info-selector"
                label="Node Data"
                value={currentNodeInfo}
                onChange={handleSelectedItemChange}
            >
                {node_options.map((key: DataInfo) => (
                    <option key={`option-${key.name}`} value={key.name}>
                        {key.label}
                    </option>
                ))}
            </NativeSelect>
        );
    }
);

NodeInfoSelector.displayName = "NodeInfoSelector";
export default NodeInfoSelector;
