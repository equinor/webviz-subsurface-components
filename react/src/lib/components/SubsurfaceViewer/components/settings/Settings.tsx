import { Icon } from "@equinor/eds-core-react";
import { layers } from "@equinor/eds-icons";
import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React, { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { MapState } from "../../redux/store";
import LayersButton from "./LayersButton";
import LayerSettingsButton from "./LayerSettingsButton";
import { getLayersInViewport } from "../../layers/utils/layerTools";

Icon.add({ layers }); // (this needs only be done once)

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: "absolute",
            bottom: theme.spacing(4),
            right: theme.spacing(2),
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
        },
    })
);

export interface SettingsProps {
    viewportId?: string;
    layerIds?: string[];
}

const Settings: React.FC<SettingsProps> = React.memo(
    ({ viewportId, layerIds }: SettingsProps) => {
        const classes = useStyles();

        const spec = useSelector((st: MapState) => st.spec);
        const [layersInView, setLayersInView] = useState<
            Record<string, unknown>[]
        >([]);
        useEffect(() => {
            const layers_in_viewport = getLayersInViewport(
                spec["layers"] as Record<string, unknown>[],
                layerIds
            ) as Record<string, unknown>[];
            setLayersInView(layers_in_viewport);
        }, [spec, layerIds]);

        if (!layersInView?.length) return null;
        return (
            <div className={classes.root}>
                {layersInView.map(
                    (layer) =>
                        layer && (
                            <LayerSettingsButton
                                layer={layer}
                                key={`layer-settings-button-${layer["id"]}-${viewportId}`}
                            />
                        )
                )}
                <LayersButton
                    id={`layers-button-${viewportId}`}
                    layers={layersInView}
                />
            </div>
        );
    }
);

Settings.displayName = "Settings";
export default Settings;
