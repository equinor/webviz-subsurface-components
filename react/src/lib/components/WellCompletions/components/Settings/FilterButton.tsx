import { Button, Icon, Tooltip } from "@equinor/eds-core-react";
import { filter_alt } from "@equinor/eds-icons";
import React, { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { updateIsDrawerOpen } from "../../redux/actions";
import { WellCompletionsState } from "../../redux/store";

// Use library approach
Icon.add({ filter_alt }); // (this needs only be done once)

/**
 * A button for toggle on and off the filter functions in the side drawer
 */
const FilterButton: React.FC = React.memo(() => {
    //Redux
    const dispatch = useDispatch();
    const isDrawerOpen = useSelector(
        (state: WellCompletionsState) => state.ui.isDrawerOpen
    );

    // Handlers
    const onClick = useCallback(
        () => dispatch(updateIsDrawerOpen(!isDrawerOpen)),
        [dispatch, isDrawerOpen]
    );

    //Render
    return (
        <div>
            <Tooltip title={isDrawerOpen ? "Close filter menu" : "Filter"}>
                <Button
                    //This is an attribute for testing purpose only.
                    //It can be removed using https://www.npmjs.com/package/babel-plugin-react-remove-properties.
                    data-testid="filter_button"
                    // Indicate the drawer is open by having the outlined border
                    variant={isDrawerOpen ? "outlined" : "ghost_icon"}
                    onClick={onClick}
                >
                    <Icon color="currentColor" name="filter_alt" />
                </Button>
            </Tooltip>
        </div>
    );
});

FilterButton.displayName = "FilterMenu";
export default FilterButton;
