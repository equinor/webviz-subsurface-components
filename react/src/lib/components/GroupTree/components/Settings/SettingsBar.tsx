import { TopBar } from "@equinor/eds-core-react";
import { createStyles, makeStyles } from "@material-ui/core";
import React from "react";
import DateTimeSlider from "./DateTimeSlider";
import FlowRateSelector from "./FlowRateSelector";
import { EdgeOptions } from "../../redux/types";

const useStyles = makeStyles(() =>
    createStyles({
        topBar: {
            minHeight: "90px",
        },
        actions: {
            position: "relative",
            display: "flex",
            flexDirection: "row",
        },
    })
);

interface Props {
    edge_options: EdgeOptions;
}

const SettingsBar: React.FC<Props> = React.memo(({ edge_options }: Props) => {
    const classes = useStyles();
    return (
        <TopBar className={classes.topBar}>
            <TopBar.Header className={classes.actions}>
                <FlowRateSelector edge_options={edge_options} />
            </TopBar.Header>
            <TopBar.Actions className={classes.actions}>
                <DateTimeSlider />
            </TopBar.Actions>
        </TopBar>
    );
});

SettingsBar.displayName = "SettingsBar";
export default SettingsBar;
