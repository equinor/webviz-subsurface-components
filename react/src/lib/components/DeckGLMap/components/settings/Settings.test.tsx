import { render } from "@testing-library/react";
import "jest-styled-components";
import React from "react";
import { Wrapper, EmptyWrapper } from "../../test/TestWrapper";
import Settings from "./Settings";

describe("test settings component", () => {
    it("snapshot test", () => {
        const { container } = render(Wrapper({ children: <Settings /> }));
        expect(container.firstChild).toMatchSnapshot();
    });
    it("test empty MapState", () => {
        const { container } = render(EmptyWrapper({ children: <Settings /> }));
        expect(container.firstChild).toMatchSnapshot();
    });
});
