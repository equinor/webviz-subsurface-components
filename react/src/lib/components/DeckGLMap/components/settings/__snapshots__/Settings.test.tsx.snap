// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test settings component snapshot test 1`] = `
@media (hover:hover) and (pointer:fine) {

}

<div
  class="makeStyles-root-1"
>
  <button
    class="MuiButtonBase-root MuiFab-root makeStyles-root-2 MuiFab-sizeMedium"
    id="hillshading-layer-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class=""
        fill="currentColor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          clip-rule="evenodd"
          d="M14 3 9.78 8.63l1.25 1.67L14 6.33 19 13h-8.46L6.53 7.63 1 15h22L14 3zM5 13l1.52-2.03L8.04 13H5zm-2 6.5L2 17h2l1 2.5H3zM8 21l-2-4h2l2 4H8zm2-4 1 2.5h2L12 17h-2zm4 0h2l2.5 5h-2L14 17zm4 0 2 4h2l-2-4h-2z"
          data-testid="eds-icon-path"
          fill-rule="evenodd"
          height="24"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiFab-root makeStyles-root-2 MuiFab-sizeMedium"
    id="wells-layer-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class=""
        fill="currentColor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          clip-rule="evenodd"
          d="M9.3 3h5.4v.9h1.8v1.8h-.9v1.717L17.784 19.2H19.2V21H4.8v-1.8h1.418L8.4 7.417V5.7h-.9V3.9h1.8V3zm4.5 2.7v1.8l.005.023-1.804 1.11-1.804-1.11.004-.023V5.7h3.6zm-3.778 2.772-.286 1.555 1.406-.865-1.12-.69zm1.979 1.218-2.484 1.529-.298 1.62L12 14.386l2.782-1.545-.298-1.621L12 9.689zm2.265.337-.286-1.555-1.12.69 1.406.865zM11.074 14.9l-2.027-1.126-.462 2.509 2.49-1.383zm.927.515 3.626 2.015.326 1.77H8.049l.325-1.77 3.627-2.015zm3.415.868L12.927 14.9l2.028-1.126.461 2.509z"
          data-testid="eds-icon-path"
          fill-rule="evenodd"
          height="24"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiFab-root makeStyles-root-2 MuiFab-sizeMedium"
    id="fault-polygons-layer-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class=""
        fill="currentColor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          clip-rule="evenodd"
          d="m18.752 10.1-2.107 3.65-2.126 3.715 4.729-2.172-1.562-.926 1.054-1.825 1.053-1.825-1.041-.617zM4.602 13.9l2.107-3.65 2.126-3.715-4.73 2.172 1.563.926-1.054 1.825-1.054 1.825 1.042.617zM11.618 5H3V3h10.353a1 1 0 0 1 .865 1.501l-7.53 13a1 1 0 0 1-.864.499H3v-2h2.247l6.371-11zM11.735 19h8.619v2H10a1 1 0 0 1-.866-1.501l7.53-13A1 1 0 0 1 17.53 6h2.823v2h-2.247l-6.37 11z"
          data-testid="eds-icon-path"
          fill-rule="evenodd"
          height="24"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiFab-root makeStyles-root-2 MuiFab-sizeMedium"
    id="drawing-layer-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class=""
        fill="currentColor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          clip-rule="evenodd"
          d="M19.169 3c-.26 0-.51.1-.71.29l-8.96 8.96 2.75 2.75 8.96-8.96a.996.996 0 0 0 0-1.41l-1.34-1.34c-.2-.2-.45-.29-.7-.29zM7.499 16c.55 0 1 .45 1 1 0 1.1-.9 2-2 2-.17 0-.33-.02-.5-.05.31-.55.5-1.21.5-1.95 0-.55.45-1 1-1zm-3 1c0-1.66 1.34-3 3-3s3 1.34 3 3c0 2.21-1.79 4-4 4-1.51 0-3.08-.78-4-2 .84 0 2-.69 2-2z"
          data-testid="eds-icon-path"
          fill-rule="evenodd"
          height="24"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <div
    id="layers-button-undefined"
  >
    <button
      class="MuiButtonBase-root MuiFab-root"
      id="layers-selector-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class=""
          fill="currentColor"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            clip-rule="evenodd"
            d="m12 16.465 7.36-5.73L21 9.465l-9-7-9 7 1.63 1.27 7.37 5.73zm-.01 2.54-7.37-5.73L3 14.535l9 7 9-7-1.63-1.27-7.38 5.74zm5.75-9.54L12 4.995l-5.74 4.47 5.74 4.47 5.74-4.47z"
            data-testid="eds-icon-path"
            fill-rule="evenodd"
            height="24"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`test settings component test empty MapState 1`] = `null`;
